### PR TITLE
Use lazy initialization for dmd.frontend

### PR DESCRIPTION
**Pull request by wilzbach**
Lazy initialization seems to be the new game in town.
`dmd.frontend` wants to play with the big kids too!

### CODE REVIEW

File: src/dmd/frontend.d [hunk 1]
```diff
@@ -27,6 +27,7 @@ This needs to be done $(I before) calling any function.
 */
 void initDMD()
 {
+    import std.concurrency : initOnce;
```
Comment (WalterBright) [inline src/dmd/frontend.d]: We're not linking with Phobos - just druntime.

**Comment by WalterBright** [comment]
I'm afraid I don't see how this improves anything.

File: src/dmd/frontend.d (hunk 1 above)
Comment (wilzbach) [reply src/dmd/frontend.d]: This file uses Phobos as it isn't part of the compiler.

**Comment by wilzbach** [comment]
The `dmd.frontend` is user-facing and this is one small step into making it easier for people to use as it reduces the number of things people can do wrong. `initOnce` is used very generously in Phobos and this file is very similar to Phobos - it's a _user_ API after all.

**Comment by jacob-carlborg** [comment]
What happens if `initDMD` is called more than once?

**Comment by rainers** [comment]
> What happens if initDMD is called more than once?

Reinitialization is not so easy due to "inaccessible" globals: here's what I had to do in my experiments for Visual D some time ago: 

https://github.com/rainers/visuald/blob/dmdserver/vdc/dmdserver.d#L1009
https://github.com/rainers/visuald/blob/dmdserver/vdc/dmdserver.d#L667

File: src/dmd/frontend.d [hunk 2]
```diff
@@ -37,16 +38,21 @@ void initDMD()
...
-    addDefaultVersionIdentifiers();
-
-    Type._init();
-    Id.initialize();
-    Module._init();
-    Target._init();
-    Expression._init();
-    Objc._init();
-    builtin_init();
+    static shared bool initialized;
```
Comment (rainers) [inline src/dmd/frontend.d]: The front end is not thread safe, no need to pretend it here.

**Comment by jacob-carlborg** [comment]
> Reinitialization is not so easy due to "inaccessible" globals: here's what I had to do in my experiments for Visual D some time ago.

I was thinking if something bad can happen if `initDMD` is called more than once.

File: src/dmd/frontend.d [hunk 3]
```diff
@@ -60,6 +66,8 @@ void addImport(string path)
     import dmd.arraytypes : Strings;
     import std.string : toStringz;
 
+    initDMD;
```
Comment (WalterBright) [inline src/dmd/frontend.d]: Please let's not introduce a new style for calling functions - append the `()`.

**Comment by WalterBright** [comment]
This just seems like over-engineering. Just use:
```
__gshared bool initialized;
assert(!initialized);
initialized = true;
```
Done - no import, no delegate, no dependencies, no "what does this do", etc.

**Pull request closed without merging**